Determine schema to document valid resource values

The CSMIM standard defines a few limited ways to specify valid resource values (respectively request payloads):

- The `enum-values` spec for resources of type enum
- The `parameters` spec for executable resources

For many resource types, additional definitions would be helpful:

- Unit of measurement for numerical resources
- Allowed value range for numerical resources
- Allowed size of arrays
- Dictionary entries, each with key and type and additional definitions as needed

In essence, this is equivalent or similar to a schema, like [JSON schema](https://json-schema.org/).

This issue is to discuss the need for and merits of such a schema for CSMIM resource values, and potential implementations in our object type YAML files.

Thanks, Uwe, for outlining various suggestions and providing detailed examples. Following our previous tagups, I’ve spent additional time reviewing options to define a schema for resource values that aligns closely with the underlying wire protocol (CBOR encoding scheme). My focus has been on finding a solution that minimally changes our existing YAML structure while expanding its expressiveness and maintaining both human readability and machine processability.

While JSON Schema offers some useful constructs, I believe we can adopt a hybrid approach that builds on our current YAML/JSON structure while leveraging Concise Data Definition Language (CDDL) as defined in [RFC 8610](https://www.rfc-editor.org/rfc/rfc8610.html). CDDL would be used selectively for defining ranges, enumerations, array sizes, and other constraints, allowing us to enhance expressiveness without requiring a complete overhaul of the schema. This approach aligns well with the goals of CSMIM and leverages existing tooling for validation.

Below, I’ve outlined my proposal along with examples to illustrate how this hybrid approach could work in practice. 

Let me know your thoughts!

## Key Objectives

The key objectives of this proposal are to enhance expressiveness with minimal disruption:

1.  **Expand Expressiveness:** Define units of measurement, value ranges, array sizes, and dictionaries with detailed key-value constraints.
2.  **Maintain Compatibility:** Ensure schema remains human-readable, machine-processable, and compatible with CBOR and JSON.
3.  **Leverage Tooling:** Benefit from existing CDDL tooling (e.g., [cddl-cat](https://github.com/hjson/cddl-cat)) to validate schemas and perform dynamic type checking.

## Proposed Enhancements

Here are the specific enhancements I propose to achieve greater expressiveness while keeping changes minimal:

### 1. Introduce New Elements in YAML

*   **Units of Measurement:** Add a `unit` field for numerical resources (e.g., `unit: ms`).
*   **Range Field:** Use a `range` field to define value ranges, enumerations, or array sizes.
*   **Optionality:** Introduce an `optional` field to explicitly mark elements as optional.

Example YAML:
  
  ```yaml
  resources:
    - id: fadeDuration
      type: uint
      range: "(0..60000)"
      description: Duration of the fade in milliseconds
      unit: ms
      optional: true
  ```
### 2. Expressiveness via CDDL Syntax

The `range` field can leverage CDDL syntax to represent:

*   **Enumerations** (e.g., `&(red: 1, green: 2)`).
*   **Numeric ranges** (e.g., `(0..65535)`).
*   **Array sizes** using occurrence counts (e.g., `[1*4]` for arrays with 1–4 elements).
*   **Complex types** like dictionaries or nested arrays.
  
  Example YAML:
  
  ```yaml
  resources:
    - id: colorCmd
      mode: w
      type: uint[]
      range: "[1*4 (0..65535)]" # Array with 1–4 elements, each in range 0–65535.
      description: Command color in CIELUV 1976 color space
    - id: colorsTest
      mode: w
      type: enum
      range: "&(black: 0, red: 1, green: 2, yellow: 3, blue: 4, magenta: 5, cyan: 6, white: 7)"
      description: Sample enum resource
  ```


### 3. Complex Type Definitions

*   Use CDDL to define complex types such as dictionaries (`map`) or nested arrays.
*   These can be documented inline or referenced in YAML.
  
Example YAML:
  
  ```yaml
  - id: schedCmd
    mode: x
    type: void
    description: Schedule fades for individual sub-units
    parameters:
      - key: c
        type: uint[][]
        range: "[+ [+ uint]]" # Defines a 2D uint array
        description: Colors for each sub-unit
        unit: "-"
  ```


  CDDL Example for Complex Types (Parameters):
  
  ```cddl
  parameters = {
    etag: bstr,
    types: [* ObjectType],
  }
  ObjectType = {
    id: tstr,
    supertypes: [* tstr],
    ? racc: bool,
  }
  ```

### 4. Tooling Benefits:
*   Adopting CDDL enables validation for both JSON and CBOR formats using existing tools (e.g., [cddl](https://www.rfc-editor.org/rfc/rfc8610.html#appendix-F), [CDDL Validator](https://cddl.one/)).
*   This ensures compliance with the schema and facilitates debugging.
  
## Complete Example (csmim.obj.light)

**Input:** YAML Schema

```yaml
---
id: csmim.obj.light.col.1
supertypes: []
description:
  A colored light unit that can consist of sub-units; Register at path
  crew/lights/units/<sec>/<nr>
deprecated: false
attributes:
  - key: sec
    description: Section number in light database
  - key: nr
    description: Consecutive unit number in section
  - key: subs
    description: Number of sub-units
    optional: false
resources:
  - id: colorCmd
    mode: w
    type: uint[]
    range: "[1*4 (0..65535)]" # Defines an array with a min of 1 to 4 elements, each containing a value within 0 to 65535.
    description: Command color in CIELUV 1976 color space [0..65535] for all sub-units
    attributes:
      - key: resattrib-test1
        description: sample resource attribute
  - id: schedCmd
    mode: x
    type: void
    description: Schedule fades for individual sub-units
    parameters:
      - key: c
        type: uint[][]
        range: "[+ [+ uint]]" # Defines a 2D uint array
        description: Colors for each sub-unit
        unit: "-"
      - key: t
        type: utc
        range: utc # Defines an utc type that is defined as either #6.1(int) or #6.1(float) per ARINC853 spec.
        description: Start time of color fade
        unit: time
      - key: d
        type: uint
        range: "(0..60000)"
        description: Fade duration
        unit: ms # Added unit for explicit reference
        optional: true
  - id: colorsTest
    mode: w
    type: enum
    range: "&( black: 0,  red: 1,  green: 2,  yellow: 3, blue: 4,  magenta: 5,  cyan: 6,  white: 7)" # An example to capture the enumerations.
    description: Sample enum resource

```

Equivalent CDDL Schema Representation

```cddl
schema-csmim.obj.light = {
  id: "csmim.obj.light.col.1",
  supertypes: [
  ],
  deprecated: false,
  resources: [
    {
      id: "colorCmd",
      mode: "w",
      type: [1*4 (0..65535)],
      resattrib-test1: tstr
    },
    {
      id: "schedCmd",
      mode: "x",
      type: void,
      parameters: {
        c: [+ [+ uint]],
        t: utc,
        ? d: (0..60000)
      }
    },
    {
      id: "colorsTest",
      mode: "w",
      type: &( black: 0,  red: 1,  green: 2,  yellow: 3, blue: 4,  magenta: 5,  cyan: 6,  white: 7)
    }
  ],
  sec: tstr,
  nr: tstr,
  subs: tstr
}
                   
```

Sample Data Validation Output (CBOR Diagnostic):

```sh
$ ./zValidateData.sh -o csmim.obj.light -f cbor-diag
Starting validation for csmim.obj.light in cbor-diag format
Validating CBOR-Diagnostic format...
[INFO] Root type for validation: schema-csmim.obj.light
[ERROR] Validation from stdin failed:
error validating at cbor location /"supertypes": expected empty array, got Array([Integer(Integer(2))])
Validation complete

$ ./zValidateData.sh -o csmim.obj.light -f cbor-diag
Starting validation for csmim.obj.light in cbor-diag format
Validating CBOR-Diagnostic format...
[INFO] Root type for validation: schema-csmim.obj.light
[ERROR] Validation from stdin failed:
error validating at cbor location /"resources"/0/"mode": expected value "w" got "wx"
Validation complete

$ ./zValidateData.sh -o csmim.obj.light -f cbor-diag
Starting validation for csmim.obj.light in cbor-diag format
Validating CBOR-Diagnostic format...
[INFO] Root type for validation: schema-csmim.obj.light
[ERROR] Validation from stdin failed:
error validating at cbor location /"resources"/2/"type": expected integer to be in range 0 <= value <= 7, got Integer(8)
Validation complete

$ ./zValidateData.sh -o csmim.obj.light -f cbor-diag
Starting validation for csmim.obj.light in cbor-diag format
Validating CBOR-Diagnostic format...
[INFO] Root type for validation: schema-csmim.obj.light
[ERROR] Validation from stdin failed:
error validating at cbor location /"resources"/2/"type": expected integer to be in range 0 <= value <= 7, got Integer(8)
error validating at cbor location : object missing key: ""subs""
Validation complete

$ ./zValidateData.sh -o csmim.obj.light -f cbor-diag
Starting validation for csmim.obj.light in cbor-diag format
Validating CBOR-Diagnostic format...
[INFO] Root type for validation: schema-csmim.obj.light
[INFO] Validation from stdin is successful
Validation complete
```

## Addressing Existing Comments

1.  **JSON Schema Incompatibilities:** I agree that standard JSON Schema differs significantly from CSMIM's structure (e.g., handling of array items and enums). Adopting JSON Schema would require substantial adaptation of our current YAML/JSON files, which may not be desirable. This approach aims to minimize that.
2.  **Superficially Similar Approach:** Continuing with our existing YAML/JSON structure while adding expressiveness through fields like `unit`, `range`, and `optional` seems like a viable path forward.
3.  **CDDL Advantages:** Unlike JSON Schema, CDDL is designed for CBOR and can describe arbitrary payloads effectively.

By selectively adopting CDDL, we enhance the schema's expressiveness while minimizing changes to our existing structure. This hybrid approach ensures compatibility and ease of adoption.

--- a/.github/schema/csmim.yaml
+++ b/.github/schema/csmim.yaml
@@ -16,34 +16,47 @@ attribute:
 resource:
   id: regex('^[A-Za-z0-9\\-._~]+$')
   mode: any('r', 'w', 'rw', 'x')
-  type: include('resource_type')
-  optional: bool(required=False)
   racc: bool(required=False)
-  description: str()
   attributes: list(include('attribute'), required=False)
+  parameters: list(include('dict_item'), required=False)
+  type: include('data_type')
+  description: str()
+  unit: str(required=False)
+  optional: bool(required=False)
   enum-values: list(include('enum_value'), required=False)
-  parameters: list(include('parameter'), required=False)
+  array-items: include('array_item', required=False)
+  dict-items: list(include('dict_item'), required=False)
 
 enum_value:
   key: int(min=0)
   name: str()
   description: str(required=False)
 
-parameter:
-  key: str()
-  type: include('resource_type')
-  optional: bool(required=False)
-  description: str()
+array_item:
+  description: str(required=False)
+  unit: str(required=False)
   enum-values: list(include('enum_value'), required=False)
+  array-items: include('array_item', required=False)
+  dict-items: list(include('dict_item'), required=False)
 
-resource_type: enum(
+dict_item:
+  key: str()
+  type: include('data_type')
+  description: str(required=False)
+  unit: str(required=False)
+  optional: bool(required=False)
+  enum-values: list(include('enum_value'), required=False)
+  array-items: include('array_item', required=False)
+  dict-items: list(include('dict_item'), required=False)
+
+data_type: enum(
   'void',
   'raw',
   'bool', 'bool[]',
   'int', 'int[]', 'int[][]',
   'uint', 'uint[]', 'uint[][]',
   'float', 'float[]', 'float[][]',
-  'string', 'string[]',
+  'string', 'string[]', 'string[][]',
   'bytes', 'bytes[]',
   'enum', 'enum[]',
   'utc', 'utc[]',

--- a/docs/schema.md
+++ b/docs/schema.md
@@ -5,15 +5,14 @@ parameters for an EXECUTE request or a response to such a request. The
 ARINC 853 standard defines in §6.2.7 facilities for specifying
 valid payloads in an object type specification.
 
-This document amends the ARINC 853 specification and provides a complete
-Schema to specify formally how a CSMIM resource value, parameter or response
-shall look like. It can be applied to nested elements, like dictionaries.
-
-
-## General structure and syntax
+This document amends the ARINC 853 specification and provides a
+Schema to specify also for nested elements, like dictionaries.
 
 CSMIM object type definitions are expressed using YAML in the CSMIM
 Knowledge Base. The CSMIM Schema therefore uses YAML, too.
+
+
+## Schema applicability
 
 There are 3 places where a CSMIM Schema occurs in an object type definition.
 Consider the following example:
@@ -26,96 +25,85 @@ resources:
   - id: value_resource
     mode: r
     type: int
-    # Put schema for resource value here
+    # Put resource schema here
     
   - id: command_resource
     mode: x
     parameters:
       - key: first_param
         type: int
-        # Put schema for parameter value here
+        # Put resource schema here
     
     type: string
-    # Put schema for return value here
+    # Put resource schema here
 ```
 
-As you can see from the above example, the schema must be put wherever a
-CSMIM data type is specified. It depends on the data type which schema
-definitions are allowed or even required. The following sections list those
-definitions for each of the data types of CSMIM.
+As you can see from the above example, the schema can be put wherever a
+CSMIM data type is specified.
 
 
-## Schema for data type `bool`
+## Resource additions
 
-There are no further definitions for boolean values:
+This amendment allows for `description` tags to add human readable descriptions.
 
-    type: bool
+For integer and floating-point numbers, you can define a `unit` for the value,
+which should be given as an SI unit.
 
+```
+type: float
+unit: km/h
+description: The speed of the device relative to the passenger
+```
 
-## Schema for the data types `int`, `uint` and `float`
+## Nested elements
 
-For integer and floating-point numbers, you can define a physical unit
-for the value, which should be given as an SI unit.
+ARINC 853 specifies `enum-values` for `enum`s. This amendment extends this pattern
+also to arrays and dictionaries:
 
-    type: float
-    unit: m/s
+```
+type: enum
+description: The description of the enum
+enum-values:
+  - name: identifier
+    key: 3
+    description: Additional notes for that key
+```
 
+This can also be applied to nested elements, for example:
 
-## Schema for the data types `string` and `bytes`
-
-There are no further definitions for string values:
-
-    type: string
-
-
-## Schema for the data type `enum`
-
-As defined in ARINC 853:
-
-    type: enum
-    enum-values:
-      - name: identifier
-        key: 3
-        description: whatever
-
-
-## Schema for the data type `utc`
-
-There are no further definitions for time stamps:
-
-    type: utc
-
-
-## Schema for the data type array
-
-For the items of the array, you can define descriptions:
-
-    type: string[]
-    array-items:
-      description: If necessary
+```
+type: enum[]
+description: If necessary
+array-items:
+  enum-values:
+    - name: identifier
+      key: 3
+      description: Additional notes for that key 
+```
 
 Note that the `type` of the array items is not specified again, because the
 array type already does so. For multidimensional arrays, you can specify the
 properties of the inner array in the first `array-items` definition, and the
 properties of the items in the second:
 
-    type: float[][]
-    array-items:
-      array-items:
-        unit: m
-
-
-## Schema for the data type `dict`
+```
+type: float[][]
+array-items:
+  array-items:
+    unit: m
+```
 
 For a dictionary, you can specify the known dictionary keys:
 
-    type: dict
-    description: A description of the dictionary
-    dict-items:
-      - key: identifier
-        optional: true
-        description: A description of the dictionary item
-        type: string
+```
+type: dict
+description: A description of the dictionary
+dict-items:
+  - key: identifier
+    optional: true
+    description: A description of the dictionary item
+    type: string
+```
 
 The `parameters` that describe the parameters of an EXECUTE request have the
 same structure as these `dict-items`.

--- a/docs/schema.md
+++ b/docs/schema.md
@@ -24,21 +24,21 @@ resources:
   - id: value_resource
     mode: r
     type: int
-    # Put schema for resource value here
+    # Put schema here to specify resource value
     
   - id: command_resource
     mode: x
     parameters:
       - key: first_param
         type: int
-        # Put schema for command parameter her
+        # Put schema here to specify command parameter
     
     type: string
-    # Put schema for response here
+    # Put schema here to specify response
 ```
 
 As you can see from the above example, the schema can be put wherever a
-CSMIM data type is specified.
+CSMIM data type is defined.
 
 
 ## Schema definitions

--- a/docs/schema.md
+++ b/docs/schema.md
@@ -1,0 +1,169 @@
+# Schema for Describing CSMIM Payloads
+
+The payload of a CSMIM message contains a CBOR-encoded resource value,
+parameters for an EXECUTE request or a responses to such a request.  The
+ARINC 853 standard already defines in §6.2.7 some facilities for specifying
+valid payloads in an object type specification.  Some elements are missing
+from the standard, though, for example a way to specify the items of a CBOR
+dictionary.
+
+This document amends the definitions of ARINC 853, thereby providing a
+complete Schema to specify formally how a CSMIM resource value, parameter or
+response may look like.
+
+The CSMIM Schema definition is much inspired by
+[JSON Schema](https://json-schema.org). It differs only slightly in
+structure and naming, in order to stay consistent with the existing
+definitions of ARINC 853.
+
+
+
+## General Structure and Syntax
+
+CSMIM object type definitions are expressed using YAML in the CSMIM
+Knowledge Base. The CSMIM Schema therefore uses YAML, too.
+
+There are 3 places where a CSMIM Schema occurs in an object type definition.
+Consider the following example:
+
+```yaml
+id: csmim.obj.something.1
+supertypes: []
+attributes: []
+resources:
+
+  - id: value_resource
+    mode: r
+    type: int
+    # Put schema for resource value here
+    
+  - id: command_resource
+    mode: x
+    parameters:
+      - key: first_param
+        type: int
+        # Put schema for parameter value here
+    
+    type: string
+    # Put schema for return value here
+```
+
+As you can see from the above example, the schema must be put wherever a
+CSMIM data type is specified. It depends on the data type which schema
+definitions are allowed or even required. The following sections list those
+definitions for each of the data types of CSMIM.
+
+
+
+## Schema for data type `bool`
+
+There are no further definitions for boolean values:
+
+    type: bool
+
+
+
+# Schema for the data types `int`, `uint` and `float`
+
+For integer and floating-point numbers, you can define the allowed range for
+the value:
+
+    type: int
+    minimum: 3
+    exclusive-minimum: 2
+    maximum: 20
+    exclusive-maximum: 21
+
+There are two types of minimum and maximum: inclusive and exclusive. The
+value must be >= the minimum respectively > the exclusive minimum. 
+
+You can also define a physical unit for the value, which must always be
+given as an SI unit.
+
+    type: float
+    unit: m/s
+
+
+
+# Schema for the data types `string` and `bytes`
+
+For human-readable strings and byte strings, you can define the allowed
+range for the string length:
+
+    type: string
+    min-length: 2
+    max-length: 3
+    
+**TODO:** Provide shortcut `length`?
+
+
+
+# Schema for the data type `enum`
+
+As defined in ARINC 853:
+
+    type: enum
+    enum-values:
+      - name: identifier
+        key: 3
+        description: whatever
+
+
+
+# Schema for the data type `utc`
+
+There are no further definitions for time stamps:
+
+    type: utc
+
+
+
+# Schema for the data type array
+
+You can define the allowed range for the number of items of an array:
+
+    type: int[]
+    min-items: 1
+    max-items: 3
+
+For the items of the array, you can define a schema that validates these:
+
+    type: string[]
+    array-items:
+      description: If necessary
+      max-length: 20
+
+Note that the `type` of the array items is not specified again, because the
+array type already does so. For multidimensional arrays, you can specify the
+properties of the inner array in the first `array-items` definition, and the
+properties of the items in the second:
+
+    type: float[][]
+    max-items: 3
+    array-items:
+      max-items: 3
+      array-items:
+        minimum: 0.0
+        maximum: 10000.0
+        unit: m
+
+
+
+# Schema for the data type `dict`
+
+For a dictionary, you can specify the known dictionary keys and a schema
+that validates their value:
+
+    type: dict
+    dict-items:
+      - key: identifier
+        optional: true
+        description: A description of the dictionary item
+        type: string
+        max-length: 8
+
+The `parameters` that describe the parameters of an EXECUTE request have the
+same structure as these `dict-items`.
+
+To ensure upwards compatibility, CSMIM dicts may always contain keys not
+specified in the schema.

--- a/docs/schema.md
+++ b/docs/schema.md
@@ -1,24 +1,16 @@
-# Schema for Describing CSMIM Payloads
+# Schema for describing CSMIM payloads
 
 The payload of a CSMIM message contains a CBOR-encoded resource value,
-parameters for an EXECUTE request or a responses to such a request.  The
-ARINC 853 standard already defines in §6.2.7 some facilities for specifying
-valid payloads in an object type specification.  Some elements are missing
-from the standard, though, for example a way to specify the items of a CBOR
-dictionary.
+parameters for an EXECUTE request or a response to such a request. The
+ARINC 853 standard defines in §6.2.7 facilities for specifying
+valid payloads in an object type specification.
 
-This document amends the definitions of ARINC 853, thereby providing a
-complete Schema to specify formally how a CSMIM resource value, parameter or
-response may look like.
-
-The CSMIM Schema definition is much inspired by
-[JSON Schema](https://json-schema.org). It differs only slightly in
-structure and naming, in order to stay consistent with the existing
-definitions of ARINC 853.
+This document amends the ARINC 853 specification and provides a complete
+Schema to specify formally how a CSMIM resource value, parameter or response
+shall look like. It can be applied to nested elements, like dictionaries.
 
 
-
-## General Structure and Syntax
+## General structure and syntax
 
 CSMIM object type definitions are expressed using YAML in the CSMIM
 Knowledge Base. The CSMIM Schema therefore uses YAML, too.
@@ -28,8 +20,7 @@ Consider the following example:
 
 ```yaml
 id: csmim.obj.something.1
-supertypes: []
-attributes: []
+
 resources:
 
   - id: value_resource
@@ -54,7 +45,6 @@ definitions are allowed or even required. The following sections list those
 definitions for each of the data types of CSMIM.
 
 
-
 ## Schema for data type `bool`
 
 There are no further definitions for boolean values:
@@ -62,43 +52,23 @@ There are no further definitions for boolean values:
     type: bool
 
 
+## Schema for the data types `int`, `uint` and `float`
 
-# Schema for the data types `int`, `uint` and `float`
-
-For integer and floating-point numbers, you can define the allowed range for
-the value:
-
-    type: int
-    minimum: 3
-    exclusive-minimum: 2
-    maximum: 20
-    exclusive-maximum: 21
-
-There are two types of minimum and maximum: inclusive and exclusive. The
-value must be >= the minimum respectively > the exclusive minimum. 
-
-You can also define a physical unit for the value, which must always be
-given as an SI unit.
+For integer and floating-point numbers, you can define a physical unit
+for the value, which should be given as an SI unit.
 
     type: float
     unit: m/s
 
 
+## Schema for the data types `string` and `bytes`
 
-# Schema for the data types `string` and `bytes`
-
-For human-readable strings and byte strings, you can define the allowed
-range for the string length:
+There are no further definitions for string values:
 
     type: string
-    min-length: 2
-    max-length: 3
-    
-**TODO:** Provide shortcut `length`?
 
 
-
-# Schema for the data type `enum`
+## Schema for the data type `enum`
 
 As defined in ARINC 853:
 
@@ -109,29 +79,20 @@ As defined in ARINC 853:
         description: whatever
 
 
-
-# Schema for the data type `utc`
+## Schema for the data type `utc`
 
 There are no further definitions for time stamps:
 
     type: utc
 
 
+## Schema for the data type array
 
-# Schema for the data type array
-
-You can define the allowed range for the number of items of an array:
-
-    type: int[]
-    min-items: 1
-    max-items: 3
-
-For the items of the array, you can define a schema that validates these:
+For the items of the array, you can define descriptions:
 
     type: string[]
     array-items:
       description: If necessary
-      max-length: 20
 
 Note that the `type` of the array items is not specified again, because the
 array type already does so. For multidimensional arrays, you can specify the
@@ -139,28 +100,22 @@ properties of the inner array in the first `array-items` definition, and the
 properties of the items in the second:
 
     type: float[][]
-    max-items: 3
     array-items:
-      max-items: 3
       array-items:
-        minimum: 0.0
-        maximum: 10000.0
         unit: m
 
 
+## Schema for the data type `dict`
 
-# Schema for the data type `dict`
-
-For a dictionary, you can specify the known dictionary keys and a schema
-that validates their value:
+For a dictionary, you can specify the known dictionary keys:
 
     type: dict
+    description: A description of the dictionary
     dict-items:
       - key: identifier
         optional: true
         description: A description of the dictionary item
         type: string
-        max-length: 8
 
 The `parameters` that describe the parameters of an EXECUTE request have the
 same structure as these `dict-items`.

--- a/docs/schema.md
+++ b/docs/schema.md
@@ -21,92 +21,145 @@ Consider the following example:
 id: csmim.obj.something.1
 
 resources:
-
   - id: value_resource
     mode: r
     type: int
-    # Put resource schema here
+    # Put schema for resource value here
     
   - id: command_resource
     mode: x
     parameters:
       - key: first_param
         type: int
-        # Put resource schema here
+        # Put schema for command parameter her
     
     type: string
-    # Put resource schema here
+    # Put schema for response here
 ```
 
 As you can see from the above example, the schema can be put wherever a
 CSMIM data type is specified.
 
 
-## Resource additions
+## Schema definitions
 
-This amendment allows for `description` tags to add human readable descriptions.
+The following sections list and explain all elements that you can use in
+CSMIM Schema definitions. This includes the elements already specified in
+ARINC 853 as well as the amendments by this document.
 
-For integer and floating-point numbers, you can define a `unit` for the value,
-which should be given as an SI unit.
 
-```
+### Descriptions
+
+You may add a `description` entry with a human-readable description at every
+position in the Schema:
+
+```yaml
 type: float
-unit: km/h
 description: The speed of the device relative to the passenger
 ```
 
-## Nested elements
 
-ARINC 853 specifies `enum-values` for `enum`s. This amendment extends this pattern
-also to arrays and dictionaries:
+### Physical units
 
+For integer and floating-point numbers, you should define a `unit` for the 
+value, if applicable. Units should belong to the SI system.
+
+```yaml
+type: float
+unit: m/s
 ```
+
+
+### Enumeration values
+
+For `enum`-typed data items, you must add an `enum-values` entry to the Schema.
+This is as specified by ARINC 853:
+
+```yaml
 type: enum
 description: The description of the enum
 enum-values:
   - name: identifier
     key: 3
-    description: Additional notes for that key
+    description: Additional notes for this enumeration value
 ```
 
-This can also be applied to nested elements, for example:
+The Schema for each enumeration value contains the following entries:
 
-```
+- `name` is a short human-readable identifier string.
+- `key` is the unsigned integer that represents the enumeration value in CBOR.
+- `description` is a longer, human-readable description (if necessary).
+
+
+### Array items
+
+For data items in an array, you may add an `array-items` entry to the Schema.
+This element contains a nested Schema that describes the array items:
+
+```yaml
 type: enum[]
-description: If necessary
+description: Description of the whole array # if necessary
 array-items:
   enum-values:
     - name: identifier
       key: 3
-      description: Additional notes for that key 
+      description: Additional notes for this enumeration value 
 ```
 
-Note that the `type` of the array items is not specified again, because the
-array type already does so. For multidimensional arrays, you can specify the
-properties of the inner array in the first `array-items` definition, and the
-properties of the items in the second:
+For multidimensional arrays, you must use nested `array-items` elements:
 
-```
+```yaml
 type: float[][]
 array-items:
   array-items:
     unit: m
 ```
 
-For a dictionary, you can specify the known dictionary keys:
+Note that the `type` of the array items is not specified again, because the
+array type already does so.
 
-```
+
+### Dictionary items
+
+For a dictionary, you should specify the known dictionary keys with a
+`dict-items` entry in the Schema:
+
+```yaml
 type: dict
 description: A description of the dictionary
 dict-items:
   - key: identifier
     optional: true
-    description: A description of the dictionary item
     type: string
+    description: A description of this dictionary item
 ```
 
-The `parameters` that describe the parameters of an EXECUTE request have the
-same structure as these `dict-items`.
+The nested Schema of a dictionary item contains the following entries:
 
-To ensure upwards compatibility, CSMIM dicts may always contain keys not
-specified in the schema.
+- `key` is the name of the item, and represents it in CBOR.
+- `optional` specifies whether the item may be omitted from the dictionary;
+  if not present, then the item is mandatory.
+- `type` is the CSMIM data type of the item.
+- all other Schema elements as applicable for the `type`.
+
+To ensure upwards compatibility, CSMIM dictionaries may always contain keys not
+specified in the Schema.
+
+
+### Parameters for commands
+
+The `parameters` Schema entry which describes the parameters of an EXECUTE 
+request has the same structure as a `dict-items` entry:
+
+```yaml
+resources:
+  - id: reset
+    mode: x
+    type: void
+    parameters:
+      - key: wait_time
+        optional: true
+        type: uint
+        unit: s
+        description: How long to wait before resetting the LRU
+```

--- a/types/csmim.obj.directory.1.yaml
+++ b/types/csmim.obj.directory.1.yaml
@@ -1,7 +1,6 @@
 %YAML 1.2
 ---
 id: csmim.obj.directory.1
-supertypes: []
 description: >
   The CSMIM resource directory service. The resource directory announces
   registration and removal of objects in the CSMIM network depending on the
@@ -18,7 +17,7 @@ description: >
   be published by the resource directory, in particular at network startup.
 
   The resource directory always resides at the path "core/csmim/directory".
-attributes: []
+
 resources:
   - id: csmim.obj
     mode: r

--- a/types/csmim.obj.registration.1.yaml
+++ b/types/csmim.obj.registration.1.yaml
@@ -34,6 +34,7 @@ resources:
             type: string
           - key: supertypes
             type: string[]
+            optional: true
           - key: resources
             type: dict[]
             dict-items:

--- a/types/csmim.obj.registration.1.yaml
+++ b/types/csmim.obj.registration.1.yaml
@@ -1,8 +1,7 @@
 %YAML 1.2
 ---
 id: csmim.obj.registration.1
-supertypes: []
-description: >-
+description: >
   CSMIM registration service interface as specified in ARINC 853.
 
   Clients should follow a certain procedure when connecting to the MQTT broker
@@ -13,12 +12,11 @@ description: >-
   "cleanSession" parameters in the MQTT CONNECT message properly, and evaluate
   the "sessionPresent" parameter in the CONNACK message it receives.
 
-attributes: []
 resources:
   - id: claim
     mode: x
     type: void
-    description: >-
+    description: >
       Requests a full registration for the CSMIM endpoint. Clients must do a
       full registration if the endpoint has not performed a successful full
       registration yet, or the registration data differs from the previous full
@@ -31,65 +29,65 @@ resources:
       - key: types
         type: dict[]
         description: List of object types used by this registration message
-        # dict-items:
-        #  - key: id
-        #    type: string
-        #  - key: supertypes
-        #    type: string[]
-        #  - key: resources
-        #    type: dict[]
-        #    dict-items:
-        #      - key: id
-        #        type: string
-        #      - key: type
-        #        type: string
-        #      - key: mode
-        #        type: string
-        #      - key: optional
-        #        type: bool
-        #        optional: true
-        #      - key: racc
-        #        type: bool
-        #        optional: true
+        dict-items:
+          - key: id
+            type: string
+          - key: supertypes
+            type: string[]
+          - key: resources
+            type: dict[]
+            dict-items:
+              - key: id
+                type: string
+              - key: type
+                type: string
+              - key: mode
+                type: string
+              - key: optional
+                type: bool
+                optional: true
+              - key: racc
+                type: bool
+                optional: true
       - key: objects
         type: dict[]
         description: List of objects to be registered
-        # dict-items:
-        #  - key: path
-        #    type: string
-        #  - key: type
-        #    type: string
-        #  - key: discoverable
-        #    type: bool
-        #    optional: true
-        #  - key: racc
-        #    type: string[][]
-        #    optional: true
-        #  - key: rmod
-        #    type: string[][]
-        #    optional: true
-        #  - key: resources
-        #    type: dict[]
-        #    dict-items:
-        #      - key: id
-        #        type: string
-        #      - key: mode
-        #        type: string
-        #        optional: true
-        #      - key: discoverable
-        #        type: bool
-        #        optional: true
-        #      - key: racc
-        #        type: string[][]
-        #        optional: true
-        #      - key: rmod
-        #        type: string[][]
-        #        optional: true
+        dict-items:
+          - key: path
+            type: string
+          - key: type
+            type: string
+          - key: discoverable
+            type: bool
+            optional: true
+          - key: racc
+            type: string[][]
+            optional: true
+          - key: rmod
+            type: string[][]
+            optional: true
+          - key: resources
+            type: dict[]
+            dict-items:
+              - key: id
+                type: string
+              - key: mode
+                type: string
+                optional: true
+              - key: discoverable
+                type: bool
+                optional: true
+              - key: racc
+                type: string[][]
+                optional: true
+              - key: rmod
+                type: string[][]
+                optional: true
 
   - id: renew
     mode: x
     type: void
-    description: >-
+    description: >
       Requests renewal of the last successful full registration for the
       endpoint. Clients should do a registration renewal if the registration
       data is unchanged from the last successful full registration and the MQTT

